### PR TITLE
Issue #403 Adding architecture diagram to Minishift usage topic

### PR DESCRIPTION
**Pull request by hferentschik**
Took a stab at a simple diagram using ditaa which is also used within the Origin docs. It ain't super pretty, but imo does a pretty good job visualizing the high-level components. Thoughts?

minishift-architecture

### CODE REVIEW

**Comment by hferentschik** [comment]
@budhrg so here we would need a new doc build image

**Comment by hferentschik** [comment]
@minishift/minishift-dev, any input on this? Regardless whether we include this in the docs or not, I'll need this for a presentation. I am interested to know whether this diagram aligns with your mental picture of Minishift and what you, for example, would change/add/remove/...

**Comment by budhrg** [comment]
@hferentschik , I like the image. However, what is the indication of three arrows from `$PATH -> minishift` ? I mean couldn't interpret it properly here.

**Comment by LalatenduMohanty** [comment]
@hferentschik The diagram looks fine as a high level diagram for users. Have you thought of putting things like `oc cluster up` , libmachine in it or you suggest another diagram which talks more about the internals.

**Comment by hferentschik** [comment]
>  I like the image. However, what is the indication of three arrows from $PATH -> minishift ? I mean couldn't interpret it properly here.

Did you read the text updates as well. I want to show that _minishift_ interacts with three different components/parts of the system - the VM, the Docker daemon running within the VM and finally the OpenShift container running on Docker. There are commands which relates to these parts (see the doc changes I am proposing as part of this diagram as well). Is it clearer now. TBH, I am also not 100% happy about the arrows yet. Maybe they should be left out?

**Comment by budhrg** [comment]
>  I am also not 100% happy about the arrows yet. Maybe they should be left out?

Yes, creating confusing IMO.

**Comment by hferentschik** [comment]
> he diagram looks fine as a high level diagram for users. 

ok, cool. A high-level view is what I am after here.

> Have you thought of putting things like oc cluster up , libmachine in it or you suggest another diagram which talks more about the internals.

Yes, but in separate diagrams. Also for a user, some of these intervals would actually be overkill. In this context I want to keep it simple.

File: docs/source/using/managing-minishift.adoc [hunk 1]
```diff
@@ -13,11 +14,47 @@ toc::[]
...
 - the OpenShift cluster provisioned by {project} within the VM
 
-The following sections contain information about managing the {project} VM.
-For details about using {project} to manage your local OpenShift cluster, see the xref:../openshift/index.adoc#[Interacting with OpenShift] section.
+The xref:../using/managing-minishift.adoc#diagram-minishift-architecture[{project} architecture] diagram outlines these components.
+The `minishift` binary, placed somewhere on `PATH`, is used to xref:../using/managing-minishift.adoc#minishift-start-overview[`start`], xref:../using/managing-minishift.adoc#minishift-stop-overview[`stop`] and xref:../using/managing-minishift.adoc#minishift-delete-overview[`delete`] the {project} VM.
+It also interacts with the Docker daemon running within the VM, for example via xref:../using/docker-daemon.adoc#reusing-docker-daemon[`minishift docker-env`].
+Finally it also communicates with the OpenShift cluster via sub-commands of the xref:../command-ref/minishift_openshift.adoc#[`minishift openshift`] command.
+
+As {project} user you interact with the OpenShift cluster using mainly the `oc` binary.
```
Comment (robin-owen) [inline docs/source/using/managing-minishift.adoc]: This sentence sounds awkward to me due to the "using mainly" -- maybe something along the lines of the following would flow better:

"As a {project} user you mainly interact with the OpenShift cluster using the `oc` binary."

File: docs/source/using/managing-minishift.adoc [hunk 2]
```diff
@@ -13,11 +14,47 @@ toc::[]
...
-- a virtual machine (VM) created by {project}
+- the {project} virtual machine (VM)
 - the OpenShift cluster provisioned by {project} within the VM
 
-The following sections contain information about managing the {project} VM.
-For details about using {project} to manage your local OpenShift cluster, see the xref:../openshift/index.adoc#[Interacting with OpenShift] section.
+The xref:../using/managing-minishift.adoc#diagram-minishift-architecture[{project} architecture] diagram outlines these components.
+The `minishift` binary, placed somewhere on `PATH`, is used to xref:../using/managing-minishift.adoc#minishift-start-overview[`start`], xref:../using/managing-minishift.adoc#minishift-stop-overview[`stop`] and xref:../using/managing-minishift.adoc#minishift-delete-overview[`delete`] the {project} VM.
+It also interacts with the Docker daemon running within the VM, for example via xref:../using/docker-daemon.adoc#reusing-docker-daemon[`minishift docker-env`].
+Finally it also communicates with the OpenShift cluster via sub-commands of the xref:../command-ref/minishift_openshift.adoc#[`minishift openshift`] command.
```
Comment (robin-owen) [inline docs/source/using/managing-minishift.adoc]: Please add a comma after "Finally".

File: docs/source/using/managing-minishift.adoc (hunk 1 above)
Comment (hferentschik) [reply docs/source/using/managing-minishift.adoc]: Sure, will fix.

File: docs/source/using/managing-minishift.adoc (hunk 2 above)
Comment (hferentschik) [reply docs/source/using/managing-minishift.adoc]: ok

**Comment by hferentschik** [comment]
I think as a first cut it might be good enough. 

@robin-owen, I applied your wording suggestions.

**Comment by LalatenduMohanty** [comment]
@hferentschik When I locally built the PR and seeing 
image and it seems the rendering did not work correctly or did I do anything wrong?

**Comment by hferentschik** [comment]
> When I locally built the PR and it seems the rendering did not work correctly or did I do anything wrong?

That's a bit vague ;-) Look at the asciidoc build logs to start with. There should be something about rendering the image. I also have no idea what image you attached. Is there a PNG file generated under the images directory or not?

**Comment by hferentschik** [comment]
@LalatenduMohanty, there you go. The message says it: "Cannot find Java executable". Do you not have any Java Runtime installed? ditaa is Java based. The plugin only contains the ditaa jar, for actual rendering you still need Java. You are on OS X, right? You should have a Java runtime.

That said, I can add a note that Java is a pre-requisite for the docs. When we sort out the image for CentOS CI, we will make sure that the in-container build works and the image is properly configured.

**Comment by hferentschik** [comment]
@LalatenduMohanty, would you mind trying again (assuming you use the incontainer build):

    $ docker rm $(docker ps --filter "status=exited" -q)
    $ docker rmi minishift/docs
    $ make clean_docs serve_docs

File: docs/source/using/managing-minishift.adoc [hunk 3]
```diff
@@ -13,11 +14,47 @@ toc::[]
...
+{project} caches this binary under `$MINISHIFT_HOME` (per default *_~/.minishift_*).
+Using xref:../openshift/openshift-client-binary.adoc#openshift-client-binary-overview[`minishift oc-env`] is an easy way to add `oc` to your `PATH`.
+For more details about using {project} to manage your local OpenShift cluster, see the xref:../openshift/index.adoc#[Interacting with OpenShift] section.
+
+[[diagram-minishift-architecture]]
+.: Minishift architecture
+[ditaa, "minishift-architecture"]
+----
+
+ localhost
```
Comment (gbraad) [inline docs/source/using/managing-minishift.adoc]: `localhost` is a networking term. i would suggest to just use `host machine`

Comment (hferentschik) [reply docs/source/using/managing-minishift.adoc]: ok

**Comment by hferentschik** [comment]
> Needs reference to:

> LiveCD ISO to make this diagram clear(er) of how the VM is created/used

I don't think this is really required on this level of abstraction. 

> persistent disk /mnt/sda1 to indicate where config and container images live

For this, I rather add another diagram to the hostfolder section. 

> but breaking this up in individual images is not preferred either

Why not? My idea is to add several diagrams as various places of the docs where the focus is on a specific detail. IMO this is preferable over cramming everything into one diagram.

**Comment by LalatenduMohanty** [comment]
@hferentschik Yes I can see the image with correct rendering now. 
screen shot 2017-10-24 at 4 24 26 pm

**Comment by hferentschik** [comment]
Rebased on master. 

@gbraad do you still want me to do additional changes?

**Comment by gbraad** [comment]
> > persistent disk /mnt/sda1 to indicate where config and container images live

> For this, I rather add another diagram to the hostfolder section.

this is not hostfolder related. :-s


I still believe the usage of the ISO belongs in this image... else, mentioning the cache in such detail for the `oc` command is also not necessary. it is about basic usage.

**Comment by hferentschik** [comment]
I am showing the cache for _oc_, not the ISO. This is relevant since _oc_ is required for interaction with OpenShift and Minishift basically provides this for you.

> this is not hostfolder related. :-s

There is is most relevant though. Otherwise it pretty much does not play a role, right? in which other context would that come in?

> I still believe the usage of the ISO belongs in this image

I might have an idea ...

**Comment by gbraad** [comment]
> > this is not hostfolder related. :-s

> There is is most relevant though. Otherwise it pretty much does not play a role, right? in which other context would that come in?

Hostfolders do not need to be mounted in the persistent disk. They just need to be mounted somewhere. There was no particular reason for me to really choose `/mnt/sda1` (besides the fact to wanting to rename this actually to `/mnt/data`).

The persistent disk is used for the storage of container images and the configuration of the OpenShift cluster.

**Comment by LalatenduMohanty** [comment]
retest this please

**Comment by LalatenduMohanty** [comment]
CI is failing with following error
```
06:57:45 - build/using/basic-usage.html
06:57:45   *  internally linking to ../using/managing-minishift.html#diagram-minishift-architecture, which does not exist (line 167)
06:57:45      <a href="../using/managing-minishift.html#diagram-minishift-architecture">Minishift architecture</a>
06:57:45   *  linking to ../using/docker-daemon.html#reusing-docker-daemon, but reusing-docker-daemon does not exist (line 169)
06:57:45      <a href="../using/docker-daemon.html#reusing-docker-daemon"></a>
06:57:45   *  trying to find hash of ../using/managing-minishift.html#diagram-minishift-architecture, but /home/docs/build/using/managing-minishift.html does not exist (line 167)
06:57:45      <a href="../using/managing-minishift.html#diagram-minishift-architecture">Minishift architecture</a>
06:57:45 htmlproofer 3.7.2 | Error:  HTML-Proofer found 3 failures!
```

File: docs/source/using/basic-usage.adoc [hunk 4]
```diff
@@ -13,11 +14,47 @@ toc::[]
 
 When you use {project}, you interact with two components:
 
-- a virtual machine (VM) created by {project}
+- the {project} virtual machine (VM)
 - the OpenShift cluster provisioned by {project} within the VM
 
-The following sections contain information about managing the {project} VM.
-For details about using {project} to manage your local OpenShift cluster, see the xref:../openshift/index.adoc#[Interacting with OpenShift] section.
+The xref:../using/managing-minishift.adoc#diagram-minishift-architecture[{project} architecture] diagram outlines these components.
```
Comment (LalatenduMohanty) [inline docs/source/using/basic-usage.adoc]: This needs to be fixed and changed to `basic-usage`.

File: docs/source/using/basic-usage.adoc [hunk 5]
```diff
@@ -13,11 +14,47 @@ toc::[]
...
+The `minishift` binary, placed somewhere on `PATH`, is used to xref:../using/basic-usage.adoc#minishift-start-overview[`start`], xref:../using/basic-usage.adoc#minishift-stop-overview[`stop`] and xref:../using/basic-usage.adoc#minishift-delete-overview[`delete`] the {project} VM.
+It also interacts with the Docker daemon running within the VM, for example via xref:../using/docker-daemon.adoc#reusing-docker-daemon[`docker-env`].
```
Comment (LalatenduMohanty) [inline docs/source/using/basic-usage.adoc]: This link also needs fix. I do not think we have `using/docker-daemon.adoc#reusing-docker-daemon` anymore.

**Comment by hferentschik** [comment]
@LalatenduMohanty broken links fixed and rebased.

**Comment by hferentschik** [comment]
I took another stab at it and worked on the diagram as well as the accompanying text. @minishift/minishift-dev have another look.

minishift-architecture

**Comment by hferentschik** [comment]
> Hostfolders do not need to be mounted in the persistent disk. They just need to be mounted somewhere. There was no particular reason for me to really choose /mnt/sda1 (besides the fact to wanting to rename this actually to /mnt/data).

The moment you want to create custom hostfolders the disk layout becomes relevant. It does not have to be /mnt/sda1, but it can. Also some of the interesting use cases are around trying to mount some of the persistent data (pvs for example) out onto the host. 
Another interesting question is around the location of the home directory which is effectively also a hostfolder. IMO the context in which the actual disk layout matters most and is most interesting is around hostfolders. 

Either way, I don't believe that the whole disk layout discussion fits into this high level diagram.

**Comment by hferentschik** [comment]
> Looks better by also giving it slightly more space

:-)

File: docs/source/using/basic-usage.adoc [hunk 6]
```diff
@@ -5,19 +5,68 @@ include::variables.adoc[]
...
+.: Minishift architecture
+[ditaa, "minishift-architecture"]
+----
+
+ host machine
+ +-------------------------------------------------------------------------------------+
+ |                      /-------------> Minishift VM                                   |
+ |                      |                                                              |
+ |                      |         /--->  o booting from pluggable Live ISO             |
+ | $PATH                |         |      o running on VirtualBox, HyperV, xhyve or KVM |
```
Comment (LalatenduMohanty) [inline docs/source/using/basic-usage.adoc]: @hferentschik I think it should be s/booting/boots/ , s/running/runs/.

Comment (hferentschik) [reply docs/source/using/basic-usage.adoc]: I think both is fine

Comment (LalatenduMohanty) [reply docs/source/using/basic-usage.adoc]: @hferentschik I think these actions are always correct i.e. truth, so these should be simple present tense. http://www.ef.com/english-resources/english-grammar/simple-present-tense/

Comment (gbraad) [reply docs/source/using/basic-usage.adoc]: The form used is called "Present Perfect Continuoust Tense" (gerund +ing) and is not really correct. But OK enough, as changes and modification will follow soon anyway

**Comment by hferentschik** [comment]
@LalatenduMohanty 

> I think it should be s/booting/boots/ , s/running/runs/.

Changed.

**Comment by LalatenduMohanty** [comment]
@hferentschik Thanks.

**Pull request merged**